pkg/mappings/mappings.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2017 by the contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mappings
+
+import (
+	"github.com/kubernetes-sigs/aws-iam-authenticator/pkg/arn"
+	"github.com/sirupsen/logrus"
+	"strings"
+	"sync"
+)
+
+// Mappings contains the IAM Identity mappings for the Authenticator to check
+// against when validating users.
+type Mappings struct {
+	lock sync.RWMutex
+
+	// Mappings is a map of IAM Identity Mappings with keys based on the rolearns
+	Identities map[string]Identity
+}
+
+// Identity is a mapping of an AWS Identities ARN to a Kubernetes username and a
+// list of Kubernetes groups. The username and groups are specified as templates
+// that may optionally contain two template parameters:
+//
+//  1) "{{AccountID}}" is the 12 digit AWS ID.
+//  2) "{{SessionName}}" is the role session name.
+//
+// The meaning of SessionName depends on the type of entity assuming the role.
+// In the case of an EC2 instance role this will be the EC2 instance ID. In the
+// case of a federated role it will be the federated identity (controlled by the
+// federated identity provider). In the case of a role assumed directly with
+// sts:AssumeRole it will be user controlled.
+//
+// You can use plain values without parameters to have a more static mapping.
+type Identity struct {
+	// ARN is the AWS Resource Name of the role. (e.g., "arn:aws:iam::000000000000:role/Foo").
+	ARN string
+
+	// Username is the username pattern that this instances assuming this
+	// role will have in Kubernetes.
+	Username string
+
+	// Groups is a list of Kubernetes groups this role will authenticate
+	// as (e.g., `system:masters`). Each group name can include placeholders.
+	Groups []string
+}
+
+// GetIdentity will return the mappings.Identity based on the ARN
+func (m *Mappings) GetIdentity(arn string) (identity Identity, ok bool) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	canonicalizedARN := getCanonicalizedARN(arn)
+	if identity, ok = m.Identities[canonicalizedARN]; ok {
+		return identity, ok
+	}
+	return identity, false
+}
+
+// AddIdentity will save the added roles to the Config
+func (m *Mappings) AddIdentity(i Identity) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	canonicalizedARN := getCanonicalizedARN(i.ARN)
+	m.Identities[canonicalizedARN] = i
+}
+
+// UpdateIdentity will do a clean delete and add of the RoleMapping
+func (m *Mappings) UpdateIdentity(oI Identity, nI Identity) {
+	m.DeleteIdentity(oI)
```

Comment:
This should have a single lock/unlock, so you can't just use the Delete/Add funcs.